aip/0002.md

```diff
@@ -11,45 +11,54 @@ redirect_from:
   - /0002
 ---
 
-# AIP Numbering
+# AIP Identifiers
 
 The AIP system provides a mechanism to index and have a single source of truth
 for API Improvement Proposals, as well as iterate on them collaboratively and
 transparently.
 
-This document describes the AIP numbering system.
+Multiple companies and teams have adopted the AIP system, with the mutual goal
+that the identifiers for common concepts should be shared between each
+company's AIP program to provide comparability.
 
 ## Assigning AIP Numbers
 
-The AIP editors (see [AIP-1](./0001.md)) are responsible for assigning a number
-to each AIP when it is accepted as a draft for review. Importantly, _all_ AIPs
-have numbers, not just approved ones. The [AIP Index](/) clearly delineates
-which AIPs are approved and binding and which are under discussion.
+The AIP editors (see [AIP A-1](./0001.md)) are responsible for assigning a
+unique identifier to each AIP when it is accepted as a draft for review. For
+AIPs expected to be common between companies, a common prefix (either one or
+two letters) and one number are assigned, and shared across the system. AIPs
+specific to a single company are given a three-letter prefix instead.
+Three-letter prefixes are only guaranteed to be unique within a given company.
 
-The editors may decide to reserve a specific block of numbers for groups of
-AIPs that are related in some way (for example, that are only scoped to a
-specific subset of APIs).
-
-Beyond this, AIP numbers are assigned arbitrarily. In general, the editors will
-take the next AIP number off of the stack to assign to a draft AIP, but
-occasionally may use a special/joke number if useful for mnemonic or other
-reasons.
+Importantly, _all_ AIPs have identifiers, not just approved ones. The
+[AIP Index](/) clearly delineates which AIPs are approved and binding and which
+are under discussion.
 
 ## AIP Blocks
 
-Currently recognized blocks of AIP numbers are:
-
-### Generally Applicable
+Common AIPs are organized into the following over-arching categories:
 
-- **1-99:** Reserved for meta-AIPs (generally process-related).
-- **100-999:** General API design guidance
+- **M:** Meta-AIPs (AIPs about AIPs)
+- **RD:** Resource design
+- **SM:** Standard methods
+- **CM:** Custom methods
+- **F:** Fields
+- **DP:** Design Patterns
+- **C:** Compatibility
+- **S:** File Structure
+- **A:** Annotations
+- **P:** Process
 
 ### Google Product Areas
 
-- **2700-2799:** Apps (GSuite)
-- **2500-2599:** Cloud
-- **3000-3099:** Actions on Google
-- **4600-4699:** Geo
+Specific Google product areas also have prefixes assigned to them. These
+prefixes are only guaranteed to be unique within Google (other companies may
+use the same prefixes in their AIP systems).
+
+- **AOG:** Actions on Google
+- **GCP:** Google Cloud
+- **GEO:** Geo
+- **GST:** GSuite
 
```

Comment:
I think this is still trying to create custom blocks for orgs or companies.

I think organizations should have the ability to extend an AIP, much like what we have been talking about with companies extending an AIP with their own domain.

Below is an example taking the custom GSuite AIP block (the only one I looked at).

1. The GSuite AIPs look like the could actually fit into the structure I proposed above using categories and sub-apis. I wrote up what that would look like below:
GSuite:
```
2712 -> Could replace G-100, or if they want to keep G-100 as it is, this could be a whole new 'G' level AIP, or depending on the actual contents of G-100, could be a sub-AIP, G-100.1

2713 -> x-ddd.1 Where x-ddd is the new identifier of AIP-213.

2715 -> Could replace G-300 or be a . extension

2716, 2717 -> Could replace or . extend F-200

2718 Might be be a brand new AIP. It could fit be a new number under an existing category, or GSuite could make their own category.
```
2. Instead of being part of the AIP identifier, the org code could be part of the path. For example, if I replaced `aip.dev/G-100` with the GSuite code above, my new AIP url would be `aip.dev/gst/G-100`. Another option might be that it is instead a query param `?org=gst`. Might be easier to work with in Jekyll, but I don't like it as much.

3. For migration purposes, redirecting `/aip-2713` to `/gst/G=100` should be pretty simple. Previously bookmarked AIPs will still work.

4. This part is just guessing, I might be wrong. I sometimes overestimate what I can do in Jekyll: I think we could configure the Jekyll pages to be dynamic enough such that we can provide a context under which the user is consuming the pages. For example, if you were in the `gst` context, and you clicked the link to `G-100`, the link would actually be to `/gst/G-100` if it existed. If it did not exist, the link would point to the default `/g-100`, but would still pass along the `gst` context so that if you clicked on an AIP that did exist in the `gst` context you would be directed to it.

Reply:
> I feel this is still trying to create custom blocks for orgs or companies.

I wrote this part under a (possibly errant assumption) that orgs and companies would still need to solve for "I want to add an AIP for this thing that _only_ applies to us".

So, I agree that the right thing for an org/company to do for an AIP within the information architecture is to extend it, but what do they do when there just is no spot in the IA at all?

Maybe the answer is, "Petition to get it added," but I sort of get the feeling we will run into "this really does not belong" at some point eventually. We could punt this sort of thing until then though.

I think another key thing on my mind here is that I expect most companies to need at least some internal-only AIPs. (We definitely do.) I think I was considering this mostly for that use case.

> Instead of being part of the AIP identifier, the org code could be part of the path. For example, if I replaced `aip.dev/G-100` with the GSuite code above, my new AIP url would be `aip.dev/gst/G-100`. Another option might be that it is instead a query param ?org=gst. Might be easier to work with in Jekyll, but I don't like it as much.

I was intending to solve for this with subdomains rather than paths under the aip.dev domain, but this is an interesting idea. I had imagined multiple (forked) repositories each serving google.aip.dev, salesforce.aip.dev, k8s.aip.dev, etc., with different people having permission on each.

Your idea might be better though. :-)

> This part is just guessing, I might be wrong. I sometimes overestimate what I can do in Jekyll: I think we could configure the Jekyll pages to be dynamic enough such that we can provide a context under which the user is consuming the pages.

Yes, you can definitely do that. I had imagined multiple repositories, and I think you might be imagining a single one, but if we go with a single one, this is definitely a thing that we could do.

> What I like about this approach, is that we are currently facing it in SF. While I am migrating our standards doc to AIPs, we have not been adding any new standards. However, our organization is currently trying to adopt the standard, and are finding organization specific exceptions and additional rules that may one day be brought into the global standard. As a result, we have the core standard doc and a type of errata. Our developers have to jump back and forth between docs looking for the discrepancies to find the source of truth. This type of approach could eliminate that.

We are on the same page here (with possibly different implementation ideas).

Reply:
> So, I agree that the right thing for an org/company to do for an AIP within the information architecture is to extend it, but what do they do when there just is no spot in the IA at all?

Oh, I don't think I called this out very well. The point I was trying to make is that best practice for an AIP adopter should be to extend or alter an existing AIP as best they can. I used GSuite's AIPs to demonstrate this, but I only vaguely call out that AIP-2718 probably does belong it's in own category. That category could be `GST`, although I think there would be some value in encouraging corps or orgs to generalize any categories they add to principles instead of their org name so that they could either be adopted later, or so that sub-orgs can extend their AIPs in a more meaningful way.

> I was intending to solve for this with subdomains rather than paths under the aip.dev domain, but this is an interesting idea. I had imagined multiple (forked) repositories each serving google.aip.dev, salesforce.aip.dev, k8s.aip.dev, etc., with different people having permission on each.

I don't think subdomains and paths are mutually exclusive ideas. Consider my real use case below with my dream implementation:

I'm on the SF API Standard team. I am also on the API standard team for my sub organization, CC.

The CC standard is based on the SF core standard. However, because CC is an acquisition, there are some legit technical boundaries that are preventing us from fully adopting the SF core standard at this time. So we have an additional set of standards that are additions or adjustments to the core standard.

So what I picture AIPs looking like would be our Salesforce core standard being under salesforce.aip.dev. Then the CC standard could provide their adjustments under a CC path.

Still haven't full thought out what this would mean for a whole new CC AIP category (CC-123), nor how or if one could extend an organizations AIPs (I can see the use case, i.e. our team or sub-org has further standards we want to enforce).

> Yes, you can definitely do that. I had imagined multiple repositories, and I think you might be imagining a single one, but if we go with a single one, this is definitely a thing that we could do.

Yeah, for sure multiple repositories for company. I actually had been envisioning multiples for orgs as well, but... yeah... after reviewing the path approach discussed above... I suppose that would imply only one.

That raises some questions. If you have one group maintaining the core repo, and another group wanting to maintain their own org specific section... they shouldn't be dependent on getting a pull request from the core team.

The answer may still be multiple repos... or, I wonder if git sub-repos might work (although I recall them being a bit of a pain to use).

> We are on the same page here (with possibly different implementation ideas).

That's the fun part, right? With lots of different ideas, we get to pick and choose from the best of them. :D

Reply:
I found myself agreeing with most of what you wrote, so I am not replying to everything.

> The CC standard is based on the SF core standard. However, because CC is an acquisition, there are some legit technical boundaries that are preventing us from fully adopting the SF core standard at this time. So we have an additional set of standards that are additions or adjustments to the core standard.
> 
> So what I picture AIPs looking like would be our Salesforce core standard being under salesforce.aip.dev. Then the CC standard could provide their adjustments under a CC path.
> 
> Still haven't full thought out what this would mean for a whole new CC AIP category (CC-123), nor how or if one could extend an organizations AIPs (I can see the use case, i.e. our team or sub-org has further standards we want to enforce).

My gut tells me the best answer here is "fork the repo", with perhaps a GitHub robot to handle sending PRs downstream to forks. Then the CC team runs their own repo and site at `cc.salesforce.aip.dev` or something similar. (DNS namespacing is awesome here.)

> That raises some questions. If you have one group maintaining the core repo, and another group wanting to maintain their own org specific section... they shouldn't be dependent on getting a pull request from the core team.

GitHub CODEOWNERS handles this really well (and we are using it now to solve this problem). It clutters the PR UI a bit but no worse.

It does mean permission is needed to set up a _new_ org-specific section, but that is not particularly onerous.

Reply:
This resolves most of my concerns until I can try it out. It makes sense.

The only other value I can see to push for a path is that it might help in the automating of the AIPs by category section. All your extensions can be listed there without needing to manually add the links to the sub-domains.

But I only say that because Jekyll and templates are my hammers, so everything looks like it can only be solved with those :D.